src/modules/audio/spatializers/oculus_spatializer.c

```diff
@@ -0,0 +1,222 @@
+    }
+
+    if (state.poseUpdated) {
+      { // Tell Oculus Audio where the headset is
+        ovrPoseStatef pose;
+
+        ma_mutex_lock(&state.poseLock); // Do nothing inside lock but make a copy of the pose
+        memcpy(&pose, &state.pose, sizeof(pose));
+        state.poseUpdated = false;
+        ma_mutex_unlock(&state.poseLock);
+
+        ovrAudio_SetListenerPoseStatef(state.context, &pose); // Upload pose
+      }
+      state.poseUpdated = false;
+    }
+  }
+
+  intptr_t *spatializerMemo = lovrSourceGetSpatializerMemoField(source);
+
+  // Lovr allows for an unlimited number of simultaneous sources but OculusAudio makes us predeclare a limit.
+  // We maintain a list of sources and keep the index each source is associated with in its memo field.
+  // So that spatializers don't need to be notified of pauses and unpauses, we assign fields anew each onPlayback call.
+  int idx = *spatializerMemo;
+
+  // This source had a record, but we gave it away.
+  if (idx >= 0 && state.sources[idx].source != source) {
+    idx = *spatializerMemo = -1;
+  }
+
+  // This source doesn't have a record. If it's playing, try to assign it one.
+  if (idx < 0 && lovrSourceIsPlaying(source)) {
+    if (state.occupiedCount < state.sourceMax) { // There's an empty slot
+      for (idx = 0; idx < state.sourceMax; idx++)
+        if (!state.sources[idx].occupied) // Claim the first unoccupied slot
+          break;
+    } else if (state.sourceCount < state.sourceMax) { // There's a slot doing a tail
+      for (idx = 0; idx < state.sourceMax; idx++)
+        if (!state.sources[idx].occupied // Claim the first unsourced slot
+         && !state.sources[idx].usedSourceThisPlayback) // Does OculusAudio allow reusing indexes within a playback? Let's guess no for now.
+          break;
+    }
+
+    if (idx >= 0) { // Successfully assigned
+      *spatializerMemo = idx;
+      state.sourceCount++;
+      state.occupiedCount++;
+      state.sources[idx].source = source;
+      state.sources[idx].occupied = true;
+      ovrAudio_ResetAudioSource(state.context, idx);
+    }
+  }
+
+  // This source has (or was just assigned) a record.
+  if (idx >= 0) {
+    uint32_t outStatus = 0;
+    state.sources[idx].usedSourceThisPlayback = true;
+
+    float position[4], orientation[4];
+    lovrSourceGetPose(source, position, orientation);
+
+    ovrAudio_SetAudioSourcePos(state.context, idx, position[0], position[1], position[2]);
+
+    ovrAudio_SpatializeMonoSourceInterleaved(state.context, idx, &outStatus, output, input);
+
+    if (!lovrSourceIsPlaying(source)) { // Source is finished
+      state.sources[idx].source = NULL;
+      *spatializerMemo = -1;
+      if (outStatus & ovrAudioSpatializationStatus_Finished) { // Source done playing, echo tailoff is done
+        state.sources[idx].occupied = false;
+      }
+    }
+    return framesOut;
+  }
+  return 0;
+}
+
+static uint32_t oculus_spatializer_tail(float *scratch, float* output, uint32_t frames) {

```

Comment:
The "scratch" buffer I added here is admittedly weird but audio.c does have an appropriately-sized buffer but sitting around.

Note that if we changed the semantics here so that instead of returning a buffer that audio.c mixes in, the tail function "transformed" the otherwise finished buffer (ie: mixed in the tails to the output itself), we could save both time (save one mix pass in audio.c) and space (scratchBuffer2 in audio.c could be half its current length because it would no longer need to store stereo). But that is less clean so I didn't want to do it without consulting…